Fix inconsistent handling for optional input parameters

MengnanLi91 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
closes #24455 

## Design
<!--A concise description (design) of the enhancement.-->
Remove the template for std::vector<T> in getParamHelper

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1707795/) on 78806ff wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25245/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1707813/) on 78806ff wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25245/coverage/xfem)

</details>



This comment will be updated on new commits.

MengnanLi91 committed:
Add detailed error message in getParam, default value in addParam, fixed framework test case #25245

Address review: add default value in addParam, fixed framework test cases

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1719555/) on ee4eb52 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 02a77588038fd83071d4a33b3ea8b1b006a02830`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1723394/) on 134f028 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a5148ca182abf18168fb36041600b93321b869f8`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1726322/) on 696c8b9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a5148ca182abf18168fb36041600b93321b869f8`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1728265/) on 2cb7c76 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a5148ca182abf18168fb36041600b93321b869f8`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1728311/) on f627bb0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a5148ca182abf18168fb36041600b93321b869f8`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1729272/) on 294fe52 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5473cf3dc207ee0c6d80ed5bf0cc4cda295a5eb8`

dschwen commented on modules/richards/src/actions/Q2PAction.C:
```diff
@@ -100,8 +100,8 @@ Q2PAction::Q2PAction(const InputParameters & params)
     _diffusivity(getParam<Real>("diffusivity")),
-    _output_nodal_masses_to(getParam<std::vector<OutputName>>("output_nodal_masses_to")),
-    _output_total_masses_to(getParam<std::vector<OutputName>>("output_total_masses_to")),
+    _output_nodal_masses_to(isParamValid("output_nodal_masses_to")? getParam<std::vector<OutputName>>("output_nodal_masses_to"):std::vector<OutputName>{}),
+    _output_total_masses_to(isParamValid("output_total_masses_to")? getParam<std::vector<OutputName>>("output_total_masses_to"):std::vector<OutputName>{}),
```
I disagree with this fix pattern. Please supply a default argument (empty vector) instead. This logic is reimplementing the default in a (for the documentation and autocomplete) intransparent way.

dschwen commented on framework/src/actions/CommonOutputAction.C:
```diff
@@ -99,10 +99,12 @@ CommonOutputAction::validParams()
 
   params.addParam<std::vector<VariableName>>(
       "hide",
+      std::vector<VariableName>{},
```
it would be cool if we could support a `std::initializer_list<T>` here for parameter type `std::vector<T>`...

MengnanLi91 replied:
`std::initializer_list<T>` could be a good idea. I fixed almost all cases in framework and modules, but the apps may be an issue here, especially I don't have access to many of them

dschwen commented on framework/src/actions/SetupMeshAction.C:
```diff
@@ -130,11 +130,9 @@ SetupMeshAction::SetupMeshAction(const InputParameters & params)
-
-  for (const auto & bnd_name : ghosted_boundaries)
-    mesh->addGhostedBoundary(mesh->getBoundaryID(bnd_name));
+  if (isParamValid("ghosted_boundaries"))
```
same here and everywhere. The right fix is to set an empty vector as a default rather than adding the logic down here

MengnanLi91 replied:
The reason to use isParamValid logic is to be more consistent with other default value check. For other data type, we allow empty values in addParam and catch them later.

dschwen replied:
That's not a good reason IMO. You are implementing the behavior that is obtained by supplying an empty vector as a default. This adds logic which we already have implemented. And the way that we have it implemented has additional functionality, namely signaling to the documentation system and language server/json dump that the parameter has a default behavior. There is no advantage to using `isParamValid` her, only drawbacks.

MengnanLi91 replied:
The default value in addParam for me is more like a suggested value for users, therefore it also shows in the documentation. A empty vector value showing there doesn't provide much useful information

dschwen replied:
Sure, it shows that the code defaults to an empty list and not to some _unknown_ behavior that is hard coded in the specific object. I don't understand why this is even controversial for all other parameter types we supply a default value when we need one. This entire PR is about making things more consistent, so why are you using a different pattern to manually implement defaults just for std::vector parameters?

MengnanLi91 replied:
These are two examples how other data type implemented(I can find more if needed). I don't know why vector parameter should all have empty default in addParam 

```
params.addParam<unsigned int>(
      "crack_front_point_index",
      "The index of the point on the crack front corresponding to this q function");
         _has_crack_front_point_index(isParamValid("crack_front_point_index")),

_crack_front_point_index(
    _has_crack_front_point_index ? getParam<unsigned int>("crack_front_point_index") : 0),
```

```
  params.addParam<std::string>("base_name", "Material property base name.");
  _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
```

MengnanLi91 replied:
This is how it looks in documentation, not some _unknown_ behavior
<img width="768" alt="Screenshot 2023-08-29 at 2 56 57 PM" src="https://github.com/idaholab/moose/assets/118846840/ba96784a-5df4-49b2-a778-4f7acc44010a">

GiudGiud replied:
Daniel is right here. It s better to have the default as it will show in the parameter as a default empty one. 
Here if base_name had a default you would see "" for the default or () or something, that would tell you more than what we have now (nothing)

GiudGiud replied:
see here

MengnanLi91 replied:
This is different from what Daniel said in my opinion.

GiudGiud replied:
@dschwen has your comment been addressed

dschwen replied:
yes

dschwen reviewed (changes requested):
Please use default values instead of isParamValid logic

dschwen commented:
add 

```
   template <typename T>
   void addParam(const std::string & name,
                 const std::initializer_list<typename T::value_type> & value,
                 const std::string & doc_string)
   {
     addParam<T>(name, T{value}, doc_string);
   }
```

in InputParameters.h

MengnanLi91 committed:
Fix modules failed tests

Add missing conductivity value to chemical_reactions failed tests

Chemical_reaction and porous_flow module fix

fix compiling error in chemical_reaction

Use _pars in chemical reaction module

Fix failed tests in reactor module and tensor_mechanics

Fix coupledvalue in tensor_mechanics

Add default value for "external_fields"

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1776632/) on 1c72a6b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 83eb82b88ae81d34bcd6e727749071fdb503f51b`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1776684/) on c4cc3cb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 83eb82b88ae81d34bcd6e727749071fdb503f51b`

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1779486/) on 574ac82 : invalidated by @MengnanLi91

mms/test.temporal FAILED (CRASH)

moosebuild commented:
Job [Docker](https://civet.inl.gov/job/1790103/) on 79a6d69 : invalidated by @MengnanLi91

Failed to find Docker

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1793718/) on c2b19b3 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.42% is less than the required 90.0%
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

MengnanLi91 committed:
Thermal_hydraulics module fix for RELAP-7

Co-Authored-By: Lise Charlot <lise.charlot@inl.gov>

dschwen commented on modules/contact/test/tests/sliding_block/in_and_out/frictional_02_penalty_out.e:
Why does this need to get regolded?

MengnanLi91 replied:
I checked the output, the difference is mainly some digital difference. The results looks the same in Paraview. I think the changes may come from the latest Petcs update like we've seen in other cases. What do you think?

dschwen replied:
Ok, let's make sure @recuero is aware of this.

recuero replied:
Note that this file is not the in `/gold` folder! It should be removed from the repo.

MengnanLi91 replied:
I double check my re-gold files and run tests on my local machine. You are right. The re-gold may not be necessary. I'll push the changes and see how the tests go

MengnanLi91 replied:
> Note that this file is not in the `/gold` folder! It should be removed from the repo.

Have deleted those two files

recuero replied:
Thanks! It was probably me (or Daniel) who mistakenly added it in the first place.

dschwen commented on modules/heat_conduction/src/dirackernels/GapHeatPointSourceMaster.C:
```diff
@@ -22,8 +22,8 @@ GapHeatPointSourceMaster::validParams()
   InputParameters params = DiracKernel::validParams();
-  params.addParam<BoundaryName>("boundary", "The primary boundary");
-  params.addParam<BoundaryName>("secondary", "The secondary boundary");
+  params.addParam<BoundaryName>("boundary", {}, "The primary boundary");
+  params.addParam<BoundaryName>("secondary", {}, "The secondary boundary");
```
This is not a vector

MengnanLi91 replied:
Good catch!

dschwen commented on modules/optimization/src/reporters/OptimizationData.C:
```diff
@@ -269,7 +269,7 @@ template <typename T>
 OptimizationDataTempl<T>::readMeasurementsFromInput()
 {
-  if (this->isParamValid("file_variable_weights"))
+  if (this->template getParam<std::vector<std::string>>("file_variable_weights").size())
```
I like to use `! .... .empty()` in these cases

dschwen commented on modules/ray_tracing/src/raykernels/RayKernelBase.C:
```diff
@@ -49,7 +49,7 @@ RayKernelBase::RayKernelBase(const InputParameters & params)
 {
   // Add dependencies
-  if (params.isParamSetByUser("depends_on"))
+  if (params.isParamSetByUser("depends_on") && isParamValid("depends_on"))
```
I think this can be just `isParamValid`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1798401/) on 82ec926 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.42% is less than the required 90.0%
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1798740/) on 6ff495b wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.42% is less than the required 90.0%
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

GiudGiud committed:
Add default param for failed tests in Modules

Add default value for "included_boundaries"

Add default value for porous_flow "phase"

Add missing clearance_schedule due to rebase

Fix high_order_elements.square_branch_tri7_2d exodiff error

Address review: minor changes, remove yaml file and modify .gitignore

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Clean format for precheck

Clean format

Clean format for pre-check

Clean format for precheck

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1798876/) on 829ef08 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.42% is less than the required 90.0%
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1840662/) on 40a8626 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 89.42% is less than the required 90.0%
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/1852060/) on f3582c0 : invalidated by @MengnanLi91

mooseutils/tests.civet_results FAILED (CRASH)

moosebuild commented:
Job [Conda (Rocky)](https://civet.inl.gov/job/1852078/) on f3582c0 : invalidated by @MengnanLi91

mooseutils/tests.civet_results FAILED (CRASH)

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1852861/) on 4dc0b1b : invalidated by @MengnanLi91

test:mortar_aux_kernels.block-dynamics-aux-fretting-wear-test-projection_angle

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1852853/) on 4dc0b1b : invalidated by @MengnanLi91

mooseutils/tests.civet_results FAILED (CRASH)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1852864/) on 4dc0b1b wanted to post the following:


The following coverage requirement(s) failed:
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1852853/) on 4dc0b1b : invalidated by @MengnanLi91

mooseutils/tests.civet_results FAILED (CRASH)

GiudGiud commented on modules/ray_tracing/src/raybcs/RayBoundaryConditionBase.C:
```diff
@@ -42,7 +42,7 @@ RayBoundaryConditionBase::RayBoundaryConditionBase(const InputParameters & param
 {
   // Add dependencies
-  if (params.isParamSetByUser("depends_on"))
+  if (params.isParamSetByUser("depends_on") && isParamValid("depends_on"))
```
```suggestion
  if (params.isParamSetByUser("depends_on"))
```

GiudGiud commented on modules/misc/misc.yaml:
```diff
@@ -0,0 +1,9 @@
+DMETHOD: opt
```
delete this file
add to .gitignore in moose/

GiudGiud commented on modules/contact/test/tests/sliding_block/in_and_out/frictional_02_penalty_out.e:
why the change?

MengnanLi91 replied:
This is not under /gold, just result exodus files

GiudGiud replied:
why are they checked in? can we generate them locally? 
I dont see what you changed that makes us have to change them

MengnanLi91 replied:
I think we can generate them locally. It doesn't break any test cases in CIVET. I though it just people forgot to delete it

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldUserObjectTransfer.C:
```diff
@@ -31,6 +31,7 @@ MultiAppGeneralFieldUserObjectTransfer::validParams()
       "Transfers user object spatial evaluations from an origin app onto a variable in the target "
       "application.");
 
+  params.set<std::vector<VariableName>>("source_variable") = std::vector<VariableName>{};
```
cant you change the default in generalFieldTransfer instead?

MengnanLi91 replied:
"source_variable" is set in 'MultiAppConservativeTransfer' and used multiple places in the code. Most of them take "source_variable" as default parameters, a few not. I think it's better to fix a few places than changing all the rest

GiudGiud commented on framework/src/meshgenerators/RenameBlockGenerator.C:
```diff
@@ -23,39 +23,11 @@ RenameBlockGenerator::validParams()
-      "name.  No change of block ID is performed, unless multiple old blocks are "
-      "given the same name, in which case they are all given the first old block "
-      "number.",
-      "Use 'new_block' instead of 'new_block_name'.");
```
this change will have to go in another PR, I m pretty sure Kylee s mesh could not be done without selecting the id independently

GiudGiud replied:
yes she has to. The new ones wont let you set ids directly

MengnanLi91 replied:
Okay, I'll put this back for this PR. I agree we need another PR. It's not good practice to rely on deprecatedParam. There should be better solution for it

GiudGiud commented on framework/src/meshgenerators/ExtraNodesetGenerator.C:
```diff
@@ -52,18 +52,6 @@ std::unique_ptr<MeshBase>
 {
   std::unique_ptr<MeshBase> mesh = std::move(_input);
-  const auto coord = getParam<std::vector<std::vector<Real>>>("coord");
-  const auto nodes = getParam<std::vector<unsigned int>>("nodes");
```
where did these checks go

GiudGiud commented on framework/src/partitioner/LibmeshPartitioner.C:
```diff
@@ -149,17 +149,18 @@ LibmeshPartitioner::prepareBlocksForSubdomainPartitioner(
   subdomain_partitioner.chunks.clear();
 
   // Insert each chunk
-  for (const auto & group : _subdomain_blocks)
```
this should have been fine with the empty default?

MengnanLi91 replied:
Review has been addressed

GiudGiud commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -56,20 +56,23 @@ FieldSplitPreconditioner::FieldSplitPreconditioner(const InputParameters & param
+    if (isParamValid("off_diag_row") && isParamValid("off_diag_column"))
+    {
+      const auto off_diag_rows = getParam<std::vector<NonlinearVariableName>>("off_diag_row");
+      const auto off_diag_columns = getParam<std::vector<NonlinearVariableName>>("off_diag_column");
```
is there a check on the size of both somewhere?

MengnanLi91 replied:
I couldn't find it. Any hints?

GiudGiud replied:
maybe add it then? seems like we would be accessing off-bounds

MengnanLi91 replied:
This review has been addressed

GiudGiud replied:
do you mind pointing me to the check?

MengnanLi91 replied:
New line 70

GiudGiud replied:
ok this is not correct.

you ll silently ignore the user inputs if off_diag_row is size 3 and off_diag_col = 0
and you ll still seg fault if
off_diag_row is size 3 and off_diag_col = 1

GiudGiud replied:
this goes beyond the PR but it d be good to fix. If you want to push this to another PR I m ok with it

MengnanLi91 replied:
check the new condition

GiudGiud replied:
you're ignoring the inputs in case they are not correct.
I think we should have an error if they are not correct.

this check should be right under if (!full)

MengnanLi91 replied:
Make an another PR for this then so that we can have proper error messages and tests for error check. I really need this one to merge quickly before any other apps make more changes and I need to patch them again. That's what happens for the Griffin test failure today.

GiudGiud replied:
other PR sounds good for this item

GiudGiud commented on framework/src/preconditioners/PhysicsBasedPreconditioner.C:
```diff
@@ -113,16 +113,20 @@ PhysicsBasedPreconditioner::PhysicsBasedPreconditioner(const InputParameters & p
+      off_diag[row].push_back(column);
+    }
   }
   // Add all of the preconditioning systems
```
I think you can move this inside the isParamValid condition

MengnanLi91 replied:
It's already been inside the isParamValid condition. That's what the line 130 `}` for

MengnanLi91 replied:
131 should stay with `addSystem(var, off_diag[var], _pre_type[var]);` Put `AddSystem` inside will cause three test cases to fail.
`problems/eigen_problem/preconditioners.newton_pbp ........................... FAILED (EXPECTED OUTPUT MISSING)
restart/restart_add_variable.add_variable_restart ............................................. FAILED (CRASH)
preconditioners/pbp.lots_of_variables ......................................................... FAILED (CRASH)
`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1857223/) on d88c0c1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 191b9de91d4081667f32371afdeb531d9f0821e2`

GiudGiud commented on modules/contact/test/tests/sliding_block/in_and_out/frictional_02_penalty_out.e:
so you just found these files and dleeted them?

MengnanLi91 replied:
This is all related to a previous discussion with Daniel and Antonio

GiudGiud commented on modules/contact/test/tests/sliding_block/in_and_out/gold/frictionless_kinematic_out.e:
why the regold?

MengnanLi91 replied:
Check this discussion https://github.com/idaholab/moose/pull/25245#discussion_r1340655402

GiudGiud replied:
this isnt a random exodus file sitting in the repo like this discussion mentions
This is a gold file, to this test
```
  [./frictionless_kinematic]
    type = 'Exodiff'
    input = 'frictionless_kinematic.i'
    exodiff = 'frictionless_kinematic_out.e'
    heavy = true
    min_parallel = 4
    abs_zero = 1e-6
    max_time = 800
    custom_cmp = 'frictionless_kinematic_out.cmp'
    issues = '#2816'
    design = 'MechanicalContactConstraint.md'
    requirement = 'We shall be able to solve the frictionless sliding block problem using a kinematic constraint formulation.'
  [../]
```

MengnanLi91 replied:
Yes,so what I did is to put it under /gold. Nothing more. I remembered before that there was a test failure. After re-gold, the test works fine. This file is not deleted but move from result folder to gold folder

MengnanLi91 replied:
Yes, this one just move from test folder to gold folder. Nothing else changes

GiudGiud replied:
It says +6 kb. Do you mind trying again but with git mv ?

GiudGiud replied:
either that or I need to learn to use exodiff to compare on my machine. Your call

MengnanLi91 replied:
Let me check

MengnanLi91 replied:
I tested both my gold file and the one copy from latest moose. Both of them passed the test. I'll just update this gold file using the one from latest moose. So there won't be 6kb changes

GiudGiud replied:
you ll need to squash this last commit into whichever commit did the change.
right now we are checking in the same file twice

GiudGiud replied:
can you also please squash in all the clang commits?

MengnanLi91 replied:
This has been addressed

moosebuild commented:
Job [Conda build (Linux)](https://civet.inl.gov/job/1857353/) on 6db02de : invalidated by @MengnanLi91

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1857387/) on 6db02de wanted to post the following:


The following coverage requirement(s) failed:
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/1859951/) on c6d00e6 : invalidated by @MengnanLi91

FAIL mooseutils/tests.civet_results FAILED (CRASH)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1859980/) on c6d00e6 wanted to post the following:


The following coverage requirement(s) failed:
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

GiudGiud commented on framework/src/materials/DerivativeSumMaterial.C:
```diff
@@ -59,7 +59,9 @@ DerivativeSumMaterialTempl<is_ad>::DerivativeSumMaterialTempl(const InputParamet
 
   // get prefactor values if not 1.0
-  std::vector<Real> p = this->template getParam<std::vector<Real>>("prefactor");
+  std::vector<Real> p = this->isParamValid("prefactor")
```
this is not needed, the param is always valid, it has a default

MengnanLi91 replied:
Review addressed

MengnanLi91 committed:
Address review: use default value when possible, replace isParamSetByUser

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Address review: remove isParamValid if necessary

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Use default {} in VariableCodensatonPreconditioner

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Address review: replace isParamSetByUser

Co-Authored-By: Daniel Schwen <daniel.schwen@inl.gov>

Clean format for precheck

Fix module tests failure

Address review: format change and remove unnecessary isParamValid check

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Address review: use size check

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1860874/) on 1363737 : invalidated by @MengnanLi91

mooseutils/tests.civet_results FAILED (CRASH)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1860885/) on 1363737 wanted to post the following:


The following coverage requirement(s) failed:
- `geochemistry` coverage rate 95.99% is less than the required 96.0%
- `level_set` coverage rate 87.36% is less than the required 88.0%
- `richards` coverage rate 93.78% is less than the required 94.0%

GiudGiud commented on framework/src/auxkernels/ParsedVectorAux.C:
```diff
@@ -98,11 +91,19 @@ ParsedVectorAux::ParsedVectorAux(const InputParameters & parameters)
+      paramError("constant_names", "Must be same length as constant_expressions");
+
+    if (isParamValid("constant_names") && isParamValid("constant_expressions"))
+      addFParserConstants(_func_F[i], constant_names[i], constant_expressions[i]);
```
why is anything changing in this file?
the previous code was fine, you could just try to use {} instead of the defaults they had

MengnanLi91 replied:
This `isParamValid` is used to replace `getParam<std::vector<std::vector<std::string>>>("constant_names").size()`. It appears multiple times in the conditions which is not necessary. `IsParamValid` is cheaper than calling 'getParam.size()'.

GiudGiud replied:
you can restore the default and then use isParamSetByUser instead of isParamValid. This lets you change a single line instead of rewriting the entire constuctor

MengnanLi91 replied:
It doesn't do any harm to the constructor or failed any test cases. Just different ways of writing it. Why this can not be accepted?

GiudGiud replied:
revert back

MengnanLi91 replied:
I don't think it is necessary to check size() multiple times

dschwen replied:
This could be made more elegant with `getParam<std::string, std::string>("constant_names", "constant_expressions")` which gets both parameters as a vector of pairs, checking that they have the same length. (with the proper change in `addFParserConstants`). Don't address this in this PR. I can make that change later.

GiudGiud commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -170,15 +170,15 @@ AdaptivityAction::act()
-    const std::vector<std::string> & weight_names =
-        getParam<std::vector<std::string>>("weight_names");
-    const std::vector<Real> & weight_values = getParam<std::vector<Real>>("weight_values");
+    if (isParamValid("weight_names") && isParamValid("weight_values"))
```
can you use a default instead? for both params

MengnanLi91 replied:
Here are test failed with default value.
`mesh/adapt.displaced_test ................................................................... FAILED (EXODIFF)
executioners/executioner.test_steady_adapt .................................................. FAILED (EXODIFF)
outputs/displaced.use_displaced ............................................................. FAILED (EXODIFF)
time_integrators/crank-nicolson.group/adapt ................................................. FAILED (EXODIFF)
mesh/adapt.test_time ........................................................................ FAILED (EXODIFF)
preconditioners/pbp.pbp/pbp_adapt_test ...................................................... FAILED (EXODIFF)
time_integrators/implicit-euler.group/adapt ................................................. FAILED (EXODIFF)
preconditioners/smp.smp/smp_adapt_test ...................................................... FAILED (EXODIFF)
time_integrators/dirk.group/dirk-2d-heat-adap ............................................... FAILED (EXODIFF)
time_integrators/bdf2.group/adapt ........................................................... FAILED (EXODIFF)
mesh/adapt.initial_adaptivity_test .......................................................... FAILED (EXODIFF)
preconditioners/reuse.mesh_refinement ....................................................... FAILED (CSVDIFF)
mesh/adapt.patch_test ....................................................................... FAILED (EXODIFF)
mesh/adapt.adapt_cycles_test ................................................................ FAILED (EXODIFF)
mesh/adapt.test ............................................................................. FAILED (EXODIFF)
bcs/periodic.testlevel1 ..................................................................... FAILED (EXODIFF)
mesh/adapt.interval ......................................................................... FAILED (EXODIFF)`

GiudGiud replied:
but why? they had a default value before.

MengnanLi91 replied:
> but why? they had a default value before. 

I think they may have 'isParamValid' check downstream. If we use default values, some of them may need change to ParamSetByUser or size check. I don't dig in because there will be a lot of tests and more(if you consider all the apps). For this PR, I just want to make sure this new changes won't break all the existing cases and code. I try not to change their logic if possible

GiudGiud replied:
They dont, I checked. 
I just did this with the {} default (e.g. reverted to what it was):
```
    // if (isParamValid("weight_names") && isParamValid("weight_values"))
    // {
    const std::vector<std::string> & weight_names =
        getParam<std::vector<std::string>>("weight_names");
    const std::vector<Real> & weight_values = getParam<std::vector<Real>>("weight_values");

    auto num_weight_names = weight_names.size();
    auto num_weight_values = weight_values.size();

    if (num_weight_names != num_weight_values)
      mooseError("Number of weight_names must be equal to number of weight_values in "
                 "Execution/Adaptivity");

    if (num_weight_names)
    {
```
and it worked fine, 0 test failures, with a default

MengnanLi91 replied:
I see, the main difference here is use default and check size vs use empty value and check isParamValid. I think both logics work under this condition. Why it has to be size check instead of isParamValid() check?

GiudGiud replied:
Because passing {}, an empty vector that has true for isParamValid (and if the user specifies "" for the parameter, it also gets true for isParamSetByUser), is not valid for the code below the if

GiudGiud commented on framework/src/auxkernels/ArrayParsedAux.C:
```diff
@@ -84,9 +84,10 @@ ArrayParsedAux::ArrayParsedAux(const InputParameters & parameters)
-  addFParserConstants(_func_F,
-                      getParam<std::vector<std::string>>("constant_names"),
-                      getParam<std::vector<std::string>>("constant_expressions"));
+  if (isParamValid("constant_names") && isParamValid("constant_expressions"))
```
try empty default instead

MengnanLi91 replied:
Do we want to go through all the `isParamValid` conditions to see if it can be replaced by default value? I've tested most of them with default value and then went back to add conditions while working on this PR. There may be a few places that it can be replaced by default value but definitely not a lot.  Feel like we are repeating all that work now and it will be very time consuming to test each changes. The current change doesn't break any test cases or apps. Could we accept it as it is?

GiudGiud replied:
It s not really a matter of seeing if you can, it s a matter that it should be possible. And if it's not we need to know why. It was certainly possible before since there was a default value. Something else must have changed

MengnanLi91 replied:
I think it is beyond the scope of this PR. There are definitely places to improve all the downstream logics but for this PR, I want to make sure all the existing cases works with this new patch. And for all the new codes there won't be any addParam with implicit default values. Otherwise I'll have to keep patching them

loganharbour replied:
is there not value in different logic for when a user provides a blank vs a user providing nothing and having a default blank?

GiudGiud replied:
yeah if we care about users doing `some_vector_param = ""` then any previous (before this PR) isParamValid logic will have to rely on the size of the vector instead.
There still wont be a need for new isParamValid logic

GiudGiud commented on framework/src/auxkernels/ParsedAux.C:
```diff
@@ -67,9 +67,10 @@ ParsedAux::ParsedAux(const InputParameters & parameters)
-  addFParserConstants(_func_F,
-                      getParam<std::vector<std::string>>("constant_names"),
-                      getParam<std::vector<std::string>>("constant_expressions"));
+  if (isParamValid("constant_names") && isParamValid("constant_expressions"))
```
same

GiudGiud commented on framework/src/meshgenerators/ExtraNodesetGenerator.C:
```diff
@@ -74,15 +62,16 @@ ExtraNodesetGenerator::generate()
-    // Our mesh may be distributed and this node may not exist on this process
-    if (!mesh->query_node_ptr(node_id))
-      continue;
+  if (isParamValid("nodes"))
```
this code would behave fine with a nodes {} default and you would not have such a large diff either

MengnanLi91 replied:
No, if you read the changing part, using 'IsValidParam' here is much better than has a bool variable flag and check.
```
 // make sure the input is not empty
  bool data_valid = false;
  if (isParamValid("nodes") && nodes.size() != 0)
    data_valid = true;
  if (_pars.isParamValid("coord") && coord.size() != 0)
    data_valid = true;
 ```

dschwen replied:
I don't get this either. Why is isParamValid necessary here? If nodes is empty the for loop is skipped automatically. I'm with Guillome. #istandwithguillaume

MengnanLi91 replied:
In the previous code, it has a mooseError to remind people the "node can not be empty". `isParamValid` will give similar error. Whoever wrote it want to have the error, I just preserve that logic instead of being silence

GiudGiud replied:
you deleted those checks and there's no size checking now

MengnanLi91 replied:
`isParamValid` gives the same thing, I don't use default value here, now if user don't provide default value, it will error out

GiudGiud commented on framework/src/meshgenerators/ParsedGenerateSideset.C:
```diff
@@ -103,25 +103,29 @@ ParsedGenerateSideset::ParsedGenerateSideset(const InputParameters & parameters)
-  addFParserConstants(_func_F,
-                      getParam<std::vector<std::string>>("constant_names"),
-                      getParam<std::vector<std::string>>("constant_expressions"));
+  if (isParamValid("constant_names") && isParamValid("constant_expressions"))
```
same

GiudGiud commented on framework/src/meshgenerators/ParsedCurveGenerator.C:
```diff
@@ -112,15 +112,18 @@ ParsedCurveGenerator::ParsedCurveGenerator(const InputParameters & parameters)
-  addFParserConstants(_func_Fz,
-                      getParam<std::vector<std::string>>("constant_names"),
-                      getParam<std::vector<std::string>>("constant_expressions"));
+  if (isParamValid("constant_names") && isParamValid("constant_expressions"))
```
same

GiudGiud commented on framework/src/meshgenerators/ParsedElementDeletionGenerator.C:
```diff
@@ -24,10 +24,11 @@ ParsedElementDeletionGenerator::validParams()
-  params.addParam<std::vector<std::string>>("constant_names",
-                                            "Vector of constants used in the parsed function");
+  params.addParam<std::vector<std::string>>(
+      "constant_names", {}, "Vector of constants used in the parsed function");
```
done right here

GiudGiud reviewed (changes requested):
Daniel's review must be addressed. Most of the added isParamValid logic is unnecessary if we just reproduced the previous behavior (the empty vector default)

MengnanLi91 commented:
> Daniel's review must be addressed. Most of the added isParamValid logic is unnecessary if we just reproduced the previous behavior (the empty vector default)

That not works everywhere. It is really case-dependent. I started with add default value but found test cases failed so I dig into the test cases and find some isParamValid logic is necessary.

GiudGiud commented:
imo the only thing that changed between before (default is silent {}) and after (actual {} in the addParam) is the behavior of isParamValid (used to be false with bad default, is true with default). Is there anything else?

MengnanLi91 commented:
> imo the only thing that changed between before (default is silent {}) and after (actual {} in the addParam) is the behavior of isParamValid (used to be false with bad default, is true with default). Is there anything else?

Check the ` AdaptivityAction` example. `IsParamValid` is a simple solution here

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1876993/) on da05036 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2f9664cc99660fa93efe982a1738e9d37ea43054`

moosebuild commented:
Job [Python 3.10 (Apptainer)](https://civet.inl.gov/job/1877065/) on d91b5ce : invalidated by @MengnanLi91

mooseutils/tests.civet_results FAILED (CRASH)

GiudGiud commented on framework/src/preconditioners/VariableCondensationPreconditioner.C:
```diff
@@ -158,21 +158,21 @@ VariableCondensationPreconditioner::VariableCondensationPreconditioner(
-    }
+    if (isParamValid("coupled_groups"))
+      for (const auto & coupled_group :
+           getParam<std::vector<NonlinearVariableName>>("coupled_groups"))
```
should be default instead of added logic

dschwen commented on framework/src/meshgenerators/ExtraNodesetGenerator.C:
```diff
@@ -91,65 +80,67 @@ ExtraNodesetGenerator::generate()
-    bool found_elem = false;
-    const Elem * elem = (*locator)(p);
-    if (elem)
+  if (_pars.isParamValid("coord"))
```
same here. Having an empty coord would also make the loop below do nothing.

MengnanLi91 replied:
This I agree but the `isParamValid("nodes")` is needed

dschwen commented on framework/src/meshgenerators/RenameBlockGenerator.C:
```diff
@@ -69,29 +73,29 @@ RenameBlockGenerator::validParams()
 RenameBlockGenerator::RenameBlockGenerator(const InputParameters & parameters)
   : MeshGenerator(parameters), _input(getMesh("input"))
 {
-  if (isParamValid("old_block_id") && isParamValid("old_block_name"))
+  if (isParamSetByUser("old_block_id") && isParamSetByUser("old_block_name"))
```
I don't understand the need for this change. Looks to me as if this code could just stay completely unchanged (no default above and continuing to use isParamValid)

MengnanLi91 replied:
It failed tests if keep unchanged

MengnanLi91 replied:
The isParamSetByUser() is removed same as all the rest below

dschwen commented on framework/src/nodalkernels/ADNodalKernel.C:
```diff
@@ -25,10 +25,10 @@ ADNodalKernel::validParams()
   : NodalKernelBase(parameters), _u(_var.adDofValues())
 {
-  if (isParamValid("save_in"))
+  if (isParamSetByUser("save_in"))
```
I don't quite trust isParamSetByUser. What happens if the param is set by an action?

dschwen replied:
As above, why not keep this file unchanged? The semantics here are still the same. If no default is provided and the user didn't set the parameter we should use `isParamValid`. This code is not relying on an implict default, so let's not change it.

MengnanLi91 replied:
Is there a better way to do it along with providing default value? Check size?

dschwen commented on framework/src/preconditioners/FieldSplitPreconditioner.C:
```diff
@@ -65,12 +66,14 @@ FieldSplitPreconditioner::FieldSplitPreconditioner(const InputParameters & param
-      unsigned int column = _nl.getVariable(0, off_diag_columns[i]).number();
-      (*cm)(row, column) = 1;
-    }
+    if (off_diag_rows.size() * off_diag_columns.size() != 0 &&
+        off_diag_rows.size() == off_diag_columns.size())
```
this should for sure be using the two parameter version of getParam instead!

dschwen commented on framework/src/preconditioners/MoosePreconditioner.C:
```diff
@@ -67,15 +70,15 @@ MoosePreconditioner::MoosePreconditioner(const InputParameters & params)
+  if (full && isParamSetByUser("off_diag_row"))
     paramError("off_diag_row", "Set full=false to specify the off-diagonal rows manually");
-  if (full && isParamValid("off_diag_column"))
+  if (full && isParamSetByUser("off_diag_column"))
```
I'll stop reviewing here and wait until we have discussed this (we can chat next week). I don't think this file should be modified either.

MengnanLi91 replied:
Sure, I can show you how it affects the test cases. Maybe there is better solution for it but simply changing back to `isParamValid` won't work

GiudGiud replied:
Probably want to check the size instead. That ended up making the most sense in a lot of these changes

dschwen reviewed (changes requested):
I don't think some of the changes in this PR should be made.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1881181/) on 74a7867 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e4e6c092437ebbd0e0e6830b6e4b87ce750ce323`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1881325/) on 4828546 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e4e6c092437ebbd0e0e6830b6e4b87ce750ce323`

MengnanLi91 commented:
@dschwen I have remove all the isParamSetByUser in the PR, if you have time, could you take a look?

GiudGiud commented on framework/src/meshgenerators/RenameBlockGenerator.C:
```diff
@@ -99,7 +99,9 @@ RenameBlockGenerator::RenameBlockGenerator(const InputParameters & parameters)
   else
   {
-    _old_block = getParam<std::vector<SubdomainName>>("old_block_name");
+    _old_block = isParamValid("old_block_name")
```
use default

MengnanLi91 replied:
Can not use default here because it will break line 81 condition and @dschwen want to avoid use `isParamSetByUser`

GiudGiud replied:
nevermind it's because it s deprecated

GiudGiud commented on framework/src/outputs/AdvancedOutput.C:
```diff
@@ -161,8 +161,9 @@ AdvancedOutput::init()
   // Separate the hide/show list into components
-  initShowHideLists(getParam<std::vector<VariableName>>("show"),
-                    getParam<std::vector<VariableName>>("hide"));
+  if (isParamValid("show") && isParamValid("hide"))
```
should use default

MengnanLi91 replied:
Good catch! This has been addressed.

GiudGiud commented on framework/src/preconditioners/MoosePreconditioner.C:
```diff
@@ -35,7 +35,8 @@ MoosePreconditioner::validParams()
   params.addParam<std::vector<NonlinearVariableName>>(
       "off_diag_column",
-      "The variable names for the off-diagonal columns you want to add into the matrix; they will "
+      "The variable names for the off-diagonal columns you want to add into the matrix; they "
+      "will "
       "be associated with an off-diagonal row from the same position in off_diag_row.");
```
```suggestion
      "The variable names for the off-diagonal columns you want to add into the matrix; they "
      "will be associated with an off-diagonal row from the same position in off_diag_row.");
```

GiudGiud commented on framework/src/scalarkernels/ParsedODEKernel.C:
```diff
@@ -69,6 +70,7 @@ ParsedODEKernel::ParsedODEKernel(const InputParameters & parameters)
   }
 
   // add postprocessors
+
```
```suggestion
```

GiudGiud commented on framework/src/transfers/MultiAppConservativeTransfer.C:
```diff
@@ -51,7 +52,9 @@ MultiAppConservativeTransfer::MultiAppConservativeTransfer(const InputParameters
     _from_postprocessors_to_be_preserved(
-        getParam<std::vector<PostprocessorName>>("from_postprocessors_to_be_preserved")),
+        _preserve_transfer
+            ? getParam<std::vector<PostprocessorName>>("from_postprocessors_to_be_preserved")
```
use a default instead

MengnanLi91 replied:
Default value can not be used here. See the line above for `_preserve_transfer `. If default value is used, `_preserve_transfer` will always be true.

GiudGiud replied:
ok that works

GiudGiud commented on modules/porous_flow/src/actions/PorousFlowActionBase.C:
```diff
@@ -99,9 +101,13 @@ PorousFlowActionBase::PorousFlowActionBase(const InputParameters & params)
-    _mass_fraction_vars(getParam<std::vector<VariableName>>("mass_fraction_vars")),
+    _mass_fraction_vars(isParamValid("mass_fraction_vars")
+                            ? getParam<std::vector<VariableName>>("mass_fraction_vars")
+                            : std::vector<VariableName>{}),
```
use default in parameter

MengnanLi91 replied:
Default can not be used here as well. See line 395, they want to error out when no value is set for "mass_fraction_vars". I don't think I should check that

GiudGiud replied:
yeah it s not a clear winner here between your code and changing the check at line 395. So let s keep that

GiudGiud commented on modules/stochastic_tools/src/trainers/GaussianProcessTrainer.C:
```diff
@@ -100,7 +100,9 @@ GaussianProcessTrainer::GaussianProcessTrainer(const InputParameters & parameter
-  std::vector<std::string> tune_parameters(getParam<std::vector<std::string>>("tune_parameters"));
+  std::vector<std::string> tune_parameters(
+      _do_tuning ? getParam<std::vector<std::string>>("tune_parameters")
+                 : std::vector<std::string>{});
```
use default in parameter

MengnanLi91 replied:
Can not use default because `_do_tuning(isParamValid("tune_parameters"))`

GiudGiud replied:
ok that works

GiudGiud commented on modules/thermal_hydraulics/src/components/HeatStructurePlate.C:
```diff
@@ -38,7 +38,8 @@ HeatStructurePlate::HeatStructurePlate(const InputParameters & params)
 
-  _material_names = getParam<std::vector<std::string>>("materials");
+  _material_names = isParamValid("materials") ? getParam<std::vector<std::string>>("materials")
+                                              : std::vector<std::string>{};
```
use default in parameter

MengnanLi91 replied:
same reason as HeatStructureCylindrical.C, can not use default here

GiudGiud replied:
see other thread

GiudGiud commented on modules/thermal_hydraulics/src/components/HeatStructureCylindrical.C:
```diff
@@ -37,7 +37,8 @@ HeatStructureCylindrical::HeatStructureCylindrical(const InputParameters & param
 
-  _material_names = getParam<std::vector<std::string>>("materials");
+  _material_names = isParamValid("materials") ? getParam<std::vector<std::string>>("materials")
+                                              : std::vector<std::string>{};
```
same

MengnanLi91 replied:
Can not used default because `checkEqualSize` will fail. The reason for it to fail is "name" is a required param. It doesn't has default value. Then the size won't be equal

GiudGiud replied:
use the size check there instead of isParamValid

MengnanLi91 replied:
I prefer not to change it. Both way works in my opinion.

GiudGiud replied:
I dont have a strong opinion there. They both work indeed and it s quite subjective

GiudGiud commented on modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C:
```diff
@@ -129,9 +129,11 @@ CrackFrontDefinition::CrackFrontDefinition(const InputParameters & parameters)
     _crack_front_points_provider(nullptr)
 {
+  auto boundary = isParamValid("boundary") ? getParam<std::vector<BoundaryName>>("boundary")
+                                           : std::vector<BoundaryName>{};
```
default in boundary restrictable could work. It might bring a lot of code to adapt though

MengnanLi91 replied:
I tried but default won't work. It failed the cases below: Mainly because "crack_front_points" and "crack_front_points_provider" uses "boundary" differently
`test:crack_tip_enrichment.crack_tip_enrichment_edge_2d ....................................... FAILED (ERRMSG)
test:solid_mechanics_basic.edge_crack_3d ..................................................... FAILED (ERRMSG)
test:solid_mechanics_basic.penny_crack ....................................................... FAILED (ERRMSG)


test:nucleation_uo.nucleateCrack ............................................................. FAILED (ERRMSG)
test:nucleation_uo.filterNucleation .......................................................... FAILED (ERRMSG)
test:solid_mechanics_basic.edge_crack_3d_fatigue ............................................. FAILED (ERRMSG)
test:solid_mechanics_basic.edge_crack_3d_mhs ................................................. FAILED (ERRMSG)
test:solid_mechanics_basic.edge_crack_2d_propagation_mhs ..................................... FAILED (ERRMSG)
test:solid_mechanics_basic.elliptical_crack .................................................. FAILED (ERRMSG)
test:solid_mechanics_basic.penny_crack_cfp ................................................... FAILED (ERRMSG)`

GiudGiud replied:
ok that was not unexpected. people have just been coding assuming no default downstream, and their code errors with 
`boundary = ""` already

GiudGiud commented on modules/porous_flow/include/actions/PorousFlowActionBase.h:
```diff
@@ -60,13 +60,13 @@ class PorousFlowActionBase : public Action, public PorousFlowDependencies
 
   /// Name of the mass-fraction variables (if any)
-  const std::vector<VariableName> & _mass_fraction_vars;
+  const std::vector<VariableName> _mass_fraction_vars;
```
you can undo that once you have the default

MengnanLi91 replied:
see the comment above

GiudGiud commented on framework/src/transfers/MultiAppDofCopyTransfer.C:
```diff
@@ -75,14 +75,14 @@ MultiAppDofCopyTransfer::initialSetup()
-                                            b))
-        paramError("from_blocks", "The block '", b, "' was not found in the mesh");
-
-    if (from_block_names.size())
```
this was handling the `from_blocks = ''` user input better.

GiudGiud replied:
that works

GiudGiud commented:
I think we should squash into a few commits now, maybe just one (the easiest option). We should avoid capturing the back & forth in the commit history on the changes. 

otherwise this is good to go

MengnanLi91 commented:
> I think we should squash into a few commits now, maybe just one (the easiest option). We should avoid capturing the back & forth in the commit history on the changes.
> 
> otherwise this is good to go

Thanks so much for all the review efforts! Agree, I'll do the clean up after @dschwen reviewed it. He still has a change request out there.

dschwen reviewed (approved):
Looks fine!

GiudGiud reviewed (changes requested):
as mentioned, this should be squashed to have either all commits contribute in the same direction or just one commit

GiudGiud reviewed (approved):
Looks good! Thanks for addressing that ticket, that was a nasty one.
